File: context/context_manager.py
import logging
import threading
import time
from typing import Dict, List, Set, Optional, Any
from threading import Timer

from agent.component import Component
from event.event import Event
from event.event_type import EventType
from event.event_dispatcher import EventDispatcher
from communicator.lapras_topic import LaprasTopic
from communicator.message_type import MessageType
from communicator.mqtt_topic import MqttTopic
from context.context_instance import ContextInstance

class ContextManager(Component):
    def __init__(self, event_dispatcher: EventDispatcher, agent):
        super().__init__(event_dispatcher, agent)
        self.logger = logging.getLogger(__name__)
        
        self.mqtt_communicator = agent.get_mqtt_communicator()
        self.context_map: Dict[str, ContextInstance] = {}
        self.contexts_published_as_updated: Set[str] = set()
        self.periodic_timers: Dict[str, Timer] = {}
    
    def subscribe_events(self):
        self.subscribe_event(EventType.MESSAGE_ARRIVED)
    
    def handle_event(self, event: Event) -> bool:
        if event.get_type() == EventType.MESSAGE_ARRIVED:
            data = event.get_data()
            topic = data[0]
            payload = data[1]
            
            if topic.get_message_type() != MessageType.CONTEXT:
                return True
            
            context_instance = ContextInstance.from_payload(payload)
            if not context_instance:
                return True
                
            self.logger.debug(
                f"Context message arrived (name={context_instance.get_name()}, "
                f"value={context_instance.get_value()}, "
                f"timestamp={context_instance.get_timestamp()}, "
                f"publisher={context_instance.get_publisher()})"
            )
            
            self.update_context_internal(context_instance)
            return True
            
        return True
    
    def publish_context(self, context_name: str, refresh_timestamp: bool = False):
        context_instance = self.context_map.get(context_name)
        if not context_instance:
            return
            
        if refresh_timestamp:
            context_instance.set_timestamp(int(time.time() * 1000))
            
        self.logger.info(f"Publishing context {context_name} = {context_instance.get_value()}")
        topic = LaprasTopic(None, MessageType.CONTEXT, context_name)
        self.mqtt_communicator.publish(topic, context_instance)
    
    def subscribe_context(self, context_name: Optional[str] = None):
        if context_name is None:
            self.logger.info("Subscribing to all contexts")
            topic = LaprasTopic(None, MessageType.CONTEXT, MqttTopic.SINGLELEVEL_WILDCARD)
        else:
            self.logger.info(f"Subscribing to context {context_name}")
            topic = LaprasTopic(None, MessageType.CONTEXT, context_name)
            
        self.mqtt_communicator.subscribe_topic(topic)
    
    def update_context(self, context_name: str, context_value: Any, publisher: str, timestamp: Optional[int] = None):
        if timestamp is None:
            timestamp = int(time.time() * 1000)
            
        context_instance = ContextInstance(context_name, context_value, timestamp, publisher)
        self.update_context_internal(context_instance)
    
    def update_context_internal(self, context_instance: ContextInstance):
        context_name = context_instance.get_name()
        former_context_instance = self.context_map.get(context_name)
        
        if (not former_context_instance or 
            former_context_instance.get_timestamp() < context_instance.get_timestamp()):
            
            self.logger.info(f"Updating context {context_name} to {context_instance.get_value()}")
            self.context_map[context_name] = context_instance
            self.dispatch_event(EventType.CONTEXT_UPDATED, context_instance)
            
            if context_name in self.contexts_published_as_updated:
                self.publish_context(context_name)
    
    def set_publish_as_updated(self, context_name: str):
        self.logger.info(f"Publish-as-updated set for context {context_name}")
        self.contexts_published_as_updated.add(context_name)
    
    def set_periodic_publish(self, context_name: str, interval: int):
        self.logger.info(f"Periodic publish set for context {context_name} every {interval} seconds")
        
        # Cancel existing timer if there is one
        if context_name in self.periodic_timers:
            self.periodic_timers[context_name].cancel()
        
        # Function that publishes and reschedules itself
        def publish_and_reschedule():
            self.publish_context(context_name, True)
            self.periodic_timers[context_name] = Timer(interval, publish_and_reschedule)
            self.periodic_timers[context_name].daemon = True
            self.periodic_timers[context_name].start()
            
        # Start the initial timer
        self.periodic_timers[context_name] = Timer(interval, publish_and_reschedule)
        self.periodic_timers[context_name].daemon = True
        self.periodic_timers[context_name].start()
    
    def get_context(self, context_name: str) -> Optional[ContextInstance]:
        return self.context_map.get(context_name)
    
    def list_contexts(self) -> List[ContextInstance]:
        return list(self.context_map.values())

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/context/ContextManager.java
package kr.ac.kaist.cdsn.lapras.context;

import kr.ac.kaist.cdsn.lapras.Agent;
import kr.ac.kaist.cdsn.lapras.Component;
import kr.ac.kaist.cdsn.lapras.communicator.LaprasTopic;
import kr.ac.kaist.cdsn.lapras.communicator.MessageType;
import kr.ac.kaist.cdsn.lapras.communicator.MqttCommunicator;
import kr.ac.kaist.cdsn.lapras.communicator.MqttTopic;
import kr.ac.kaist.cdsn.lapras.event.Event;
import kr.ac.kaist.cdsn.lapras.event.EventDispatcher;
import kr.ac.kaist.cdsn.lapras.event.EventType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.*;

/**
 * Created by Daekeun Lee on 2016-11-14.
 */
public class ContextManager extends Component {
    private static Logger LOGGER = LoggerFactory.getLogger(ContextManager.class);

    private final ScheduledExecutorService periodicPublisher = Executors.newScheduledThreadPool(1);
    private final MqttCommunicator mqttCommunicator;
    private final ConcurrentMap<String, ContextInstance> contextMap = new ConcurrentHashMap<>();
    private final Set<String> contextsPublishedAsUpdated = new CopyOnWriteArraySet<>();

    public ContextManager(EventDispatcher eventDispatcher, Agent agent) {
        super(eventDispatcher, agent);
        this.mqttCommunicator = agent.getMqttCommunicator();
    }

    @Override
    protected void subscribeEvents() {
        subscribeEvent(EventType.MESSAGE_ARRIVED);
    }

    @Override
    protected boolean handleEvent(Event event) {
        switch (event.getType()) {
            case MESSAGE_ARRIVED:
                LaprasTopic topic = (LaprasTopic) ((Object[])event.getData())[0];
                byte[] payload = (byte[]) ((Object[])event.getData())[1];
                if(!topic.getMessageType().equals(MessageType.CONTEXT)) break;
                ContextInstance contextInstance = ContextInstance.fromPayload(payload);
                if(contextInstance == null) break;

                LOGGER.debug("Context message arrived (name={}, value={}, timestamp={}, publisher={})",
                        contextInstance.getName(), contextInstance.getValue(), contextInstance.getTimestamp(), contextInstance.getPublisher());
                updateContext(contextInstance);
                return true;
        }
        return true;
    }

    public void publishContext(String contextName, boolean refreshTimestamp) {
        ContextInstance contextInstance = contextMap.get(contextName);
        if(contextInstance == null) return;

        if(refreshTimestamp) contextInstance.setTimestamp(System.currentTimeMillis());

        LOGGER.info("Publishing contextInstance {} = {}", contextName, contextInstance.getValue());
        LaprasTopic topic = new LaprasTopic(null, MessageType.CONTEXT, contextName);
        mqttCommunicator.publish(topic, contextInstance);
    }

    public void publishContext(String contextName) {
        publishContext(contextName, false);
    }

    public void subscribeContext(String contextName) {
        LaprasTopic topic;
        if (contextName == null) {
            LOGGER.info("Subscribing to all contexts");
            topic = new LaprasTopic(null, MessageType.CONTEXT, MqttTopic.SINGLELEVEL_WILDCARD);
        } else {
            LOGGER.info("Subscribing to context {}", contextName);
            topic = new LaprasTopic(null, MessageType.CONTEXT, contextName);
        }
        mqttCommunicator.subscribeTopic(topic);
    }

    public void updateContext(String contextName, Object contextValue, String publisher) {
        long timestamp = System.currentTimeMillis(); // Assuming that all devices are set to the same timezone
        ContextInstance contextInstance = new ContextInstance(contextName, contextValue, timestamp, publisher);
        updateContext(contextInstance);
    }

    public void updateContext(String contextName, Object contextValue, String publisher, Long timestamp) {
        ContextInstance contextInstance = new ContextInstance(contextName, contextValue, timestamp, publisher);
        updateContext(contextInstance);
    }

    private void updateContext(ContextInstance contextInstance) {
        ContextInstance formerContextInstance = contextMap.get(contextInstance.getName());
        if(formerContextInstance == null || formerContextInstance.getTimestamp() < contextInstance.getTimestamp()) {
            LOGGER.info("Updating contextInstance {} to {}", contextInstance.getName(), contextInstance.getValue());
            contextMap.put(contextInstance.getName(), contextInstance);
            dispatchEvent(EventType.CONTEXT_UPDATED, contextInstance);

            if(contextsPublishedAsUpdated.contains(contextInstance.getName())) {
                publishContext(contextInstance.getName());
            }
        }
    }

    public void setPublishAsUpdated(String contextName) {
        LOGGER.info("Publish-as-updated set for context {}", contextName);
        contextsPublishedAsUpdated.add(contextName);
    }

    public void setPeriodicPublish(String contextName, Integer interval) {
        LOGGER.info("Periodic publish set for context {} every {} seconds", contextName, interval);
        periodicPublisher.scheduleAtFixedRate(()->publishContext(contextName, true), 0, interval, TimeUnit.SECONDS);
    }

    public ContextInstance getContext(String contextName) {
        return contextMap.get(contextName);
    }

    public List<ContextInstance> listContexts() {
        return new ArrayList<>(contextMap.values());
    }
}
